compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionPreprocessing.cpp

```diff
@@ -131,14 +141,63 @@ struct FoldSuccessiveTensorInsertSliceOps
   }
 };
 
+struct GatherFusionPattern : public OpRewritePattern<linalg::YieldOp> {
+  using OpRewritePattern<linalg::YieldOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::YieldOp yieldOp,
+                                PatternRewriter &rewriter) const override {
+    // Specific case. The linalg generic implementation of "gather"
+    // cannot be fused because it there is no producer-consumer
+    // relationship between the two generics. This is because the indexing
+    // is not affine (index values come from a tensor).
+    if (yieldOp->getNumOperands() != 1) {
+      return failure();
+    }
+    auto extractOp = dyn_cast_or_null<tensor::ExtractOp>(
+        yieldOp->getOperand(0).getDefiningOp());
+    if (!extractOp) {
+      return failure();
+    }
+
+    // match the generic that gens a higher bitwidth tensor
+    auto definingOp = dyn_cast_or_null<linalg::GenericOp>(
+        extractOp.getOperand(0).getDefiningOp());
+    if (!definingOp) {
+      return rewriter.notifyMatchFailure(
+          yieldOp, "expected extract operand to be a generic op");
+    }
+
+    // generic body should contain only a arith.extf and a linalg.yield
+    auto &ops = definingOp->getRegion(0).front().getOperations();
+    if (ops.size() != 2 || !isa<arith::ExtFOp>(ops.front()) ||
+        !isa<linalg::YieldOp>(ops.back())) {
+      return rewriter.notifyMatchFailure(yieldOp,
+                                         "expected generic op to have 2 ops");
+    }
```

Comment:
Maybe this is a better way to do it (more generalized):

```cpp
// Constraints for defining op:
// 1. Must be a linalg.generic
// 2. Must only have one input & output
// 3. Must be elementwise & identity
// 3. Result is only used by extract
if (!definingOp || definingOp.getNumDpsInputs() != 1 ||
    definingOp->getNumResults() != 1 ||
    !linalg::isElementwise(definingOp) ||
    llvm::any_of(definingOp.getIndexingMapsArray(),
                 [](auto &map) { return !map.isIdentity(); }) ||
    llvm::any_of(
        definingOp.getResult(0).getUsers(),
        [&extractOp](const auto &user) { return user != extractOp; })) {
  return failure();
}
```


It seems like any operation that is effectively a map over the tensor could be fused, not just extfop

Reply:
You could just use the isDequantization function defined in RegionOpUtils.h to check for a dequantization operation.

Reply:
You can also just ignore the number of uses of the producer. If there are multiple users, they will handled in their own appropriate way

Reply:
Actually you probably need to also check that there is only one input and one output of the defining op. Rest of the constraints on `isDequantizationOp` should be enough.